packages/gatsby-plugin-fastify/src/gatsby-node.ts

```diff
@@ -7,23 +8,60 @@ import type { GatsbyNode } from "gatsby";
 
 import { makePluginData } from "./utils/plugin-data";
 import { getFunctionManifest } from "./gatsby/funcitons-manifest";
-import { CONFIG_FILE_NAME } from "./utils/constants";
+import { CONFIG_FILE_NAME, BUILD_HTML_STAGE, BUILD_CSS_STAGE } from "./utils/constants";
 import { getClientSideRoutes } from "./gatsby/client-side-route";
 import { getServerSideRoutes } from "./gatsby/server-routes";
 import { getProxiesAndRedirects } from "./gatsby/proxies-and-redirects";
+import { buildHeadersProgram, Page } from "./gatsby/header-builder";
+
+const assetsManifest: WebpackAssetsManifest.Assets = {};
+// Inject a webpack plugin to get the file manifests so we can translate all link headers
+export const onCreateWebpackConfig: GatsbyNode["onCreateWebpackConfig"] = ({ actions, stage }) => {
+  //@ts-expect-error Gatsby seems to be miss typed
+  if (stage !== BUILD_HTML_STAGE && stage !== BUILD_CSS_STAGE) {
+    return;
+  }
+
+  actions.setWebpackConfig({
+    plugins: [
+      new WebpackAssetsManifest({
+        assets: assetsManifest, // mutates object with entries
+        merge: true,
+      }),
+    ],
+  });
+};
 
 export const onPostBuild: GatsbyNode["onPostBuild"] = async (
   { store, pathPrefix, reporter },
   pluginOptions: GatsbyFastifyPluginOptions
 ) => {
   try {
     const { proxies, redirects } = getProxiesAndRedirects(store);
-    const pluginData = await makePluginData(store, pathPrefix);
+    const pluginData = await makePluginData(store, assetsManifest, pathPrefix);
 
     const functions = await getFunctionManifest(pluginData);
     const clientSideRoutes = await getClientSideRoutes(pluginData);
     const serverSideRoutes = await getServerSideRoutes(pluginData);
 
+    // combine all pages
+    const pages: Array<Page> = [
+      ...clientSideRoutes.map(({ matchPath: path }): Page => ({ path: path, mode: "CSR" })),
+      ...serverSideRoutes.map(({ path, mode }): Page => ({ path, mode })),
+      ...redirects.map(({ fromPath: path }): Page => ({ path, mode: "REDIRECT" })),
+      ...proxies.map(({ fromPath: path }): Page => ({ path, mode: "PROXY" })),
+      ...functions.map(
+        ({ matchPath: path, functionRoute }): Page => ({
+          path: path ?? functionRoute,
+          mode: "FUNCTION",
+        })
+      ),
+      ...[...pluginData.pages].map(([, value]) => ({ path: value.path, mode: value.mode })),
```

Comment:
What's up with this double expansion of the array?

Reply:
ESLint did that during `lint-staged`. It was originally `...Array.from(pluginData.pages).map`. `pluginData.pages` is of type `Map` which when you expand it returns an array of arrays with `[key, value]`: 

```js
[
  [0: "/path", 1: { mode: "SSG", path: "/path" }]
]
```

which we then `.map` over to get the array of objects we want and then we expand that into pages with the rest of them.